Use of RPATH for implicit link directories problematic for build systems that use `LD_LIBRARY_PATH=<build dir> ./executable`

### Steps to reproduce this issue:

1. Use a compiler that has its runtime libraries in a directory that also contains `lib<example>.so`
2. Install the package `spack install <example>`
3. The build system invokes its just-built executable: `LD_LIBRARY_PATH=<build dir> ./example ...`

The latter fails because the `example` executable contains an RPATH pointing to the implicit link directory of the compiler, which happens to contain `libexample.so` -- this library is picked up instead of the just-built `libexample.so` in `LD_LIBRARY_PATH`.

Notice that Spack itself adds the rpath, the system compiler would only inject the `-L` flag.

The pattern `LD_LIBRARY_PATH=. ./example` is probably most common in autotools packages. CMake packages are not affected as they use build rpaths to work around this problem.

Below is an example where this happens with `perl`.



## Original issue


### Steps to reproduce the issue

```console
[csbcfmpd@login209-18 spack-src]$ spack spec -I perl
Input spec
--------------------------------
 -   perl

Concretized
--------------------------------
 -   perl@5.38.0%gcc@10.1.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-rocky8-icelake
[+]      ^berkeley-db@18.1.40%gcc@10.1.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-rocky8-icelake
[+]          ^gmake@4.4.1%gcc@10.1.0~guile build_system=generic arch=linux-rocky8-icelake
[+]      ^bzip2@1.0.8%gcc@10.1.0~debug~pic+shared build_system=generic arch=linux-rocky8-icelake
[+]          ^diffutils@3.9%gcc@10.1.0 build_system=autotools arch=linux-rocky8-icelake
[+]              ^libiconv@1.17%gcc@10.1.0 build_system=autotools libs=shared,static arch=linux-rocky8-icelake
[+]      ^gdbm@1.23%gcc@10.1.0 build_system=autotools arch=linux-rocky8-icelake
[+]          ^readline@8.2%gcc@10.1.0 build_system=autotools patches=bbf97f1 arch=linux-rocky8-icelake
[+]              ^ncurses@6.4%gcc@10.1.0~symlinks+termlib abi=none build_system=autotools arch=linux-rocky8-icelake
[+]                  ^pkgconf@1.9.5%gcc@10.1.0 build_system=autotools arch=linux-rocky8-icelake
[+]      ^zlib-ng@2.1.4%gcc@10.1.0+compat+opt build_system=autotools arch=linux-rocky8-icelake
```


### Error message

<details><summary>loadable library and perl binaries are mismatched</summary>
[csbcfmpd@login209-18 spack-src]$ spack install -j 40  perl
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/berkeley-db-18.1.40-i3j6vwjja62njh2mi6dgg5mqel3m6ani
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/bzip2-1.0.8-oo6spantejf27dmb7bzwccfttwtfx7bb
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/ncurses-6.4-2kenxblld7jmiylsjweggbentnz3yrbm
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/zlib-ng-2.1.4-575jc3ailnn2g7unxmoivbsov7vls7sw
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/readline-8.2-cpx6vq4aolpyqj7tguzpiha4hpuexnr5
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/gdbm-1.23-ei62qyttjgwe6qlz3sgbipffqt6b7ome
==> Installing perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6 [7/7]
==> No binary for perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6 found: installing from source
==> Using cached archive: /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack/var/spack/cache/_source-cache/archive/21/213ef58089d2f2c972ea353517dc60ec3656f050dcc027666e118b508423e517.tar.gz
==> Using cached archive: /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack/var/spack/cache/_source-cache/archive/9d/9da50e155df72bce55cb69f51f1dbb4b62d23740fb99f6178bb27f22ebdf8a46.tar.gz
==> Moving resource stage
        source: /tmp/csbcfmpd/spack-stage/resource-cpanm-2tubazheh7m4pdei3eugarkkduubhga6/spack-src/
        destination: /tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src/cpanm/cpanm
==> Applied patch /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack/var/spack/repos/builtin/packages/perl/zlib-ng.patch
==> Ran patch() for perl
==> perl: Executing phase: 'configure'
==> perl: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j40'

1 error found in build log:
     2831          \
     2832    
     2833    chmod 755 ../../lib/auto/Unicode/Normalize/Normalize.so
     2834    make[1]: Leaving directory '/tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src/dist/Unicode-Normalize'
     2835    LD_LIBRARY_PATH=/tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src  ./perl -Ilib -I. -f pod/buildtoc -q
     2836    Cwd.c: loadable library and perl binaries are mismatched (got handshake key 0xf380080, needed 0xcd00080)
  >> 2837    make: *** [makefile:426: pod/perltoc.pod] Error 1

See build log for details:
  /tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-build-out.txt
</details>


### Information on your system

* **Spack:** 0.21.0.dev0 (80944d22f775b014cfd201e895fc8777a2038786)
* **Python:** 3.7.8
* **Platform:** linux-rocky8-icelake
* **Concretizer:** clingo


### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/13244340/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/13244341/spack-build-out.txt)

The problem is that the perl version compiled in spack-src is not the 5.38.0 but the 5.30.3
```
[csbcfmpd@login209-18 ~]$ cd /tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src
[csbcfmpd@login209-18 spack-src]$ ./perl --version
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
        LANGUAGE = (unset),
        LC_ALL = (unset),
        LC_MEASUREMENT = "de_DE.UTF-8",
        LC_PAPER = "de_DE.UTF-8",
        LC_MONETARY = "de_DE.UTF-8",
        LC_NAME = "de_DE.UTF-8",
        LC_ADDRESS = "de_DE.UTF-8",
        LC_NUMERIC = "de_DE.UTF-8",
        LC_TELEPHONE = "de_DE.UTF-8",
        LC_IDENTIFICATION = "de_DE.UTF-8",
        LC_TIME = "de_DE.UTF-8",
        LANG = "en_US.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to a fallback locale ("en_US.UTF-8").

This is perl 5, version 30, subversion 3 (v5.30.3) built for x86_64-linux-thread-multi
(with 30 registered patches, see perl -V for more detail)

Copyright 1987-2020, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
```
maintainer: @LydDeb

The issue is related with:
https://github.com/spack/spack/issues/26738

A direct workaround is to install perl@5.30.3

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

same
```
==> Installing perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6 [7/7]
==> No binary for perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6 found: installing from source
==> Using cached archive: /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack/var/spack/cache/_source-cache/archive/21/213ef58089d2f2c972ea353517dc60ec3656f050dcc027666e118b508423e517.tar.gz
==> Using cached archive: /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack/var/spack/cache/_source-cache/archive/9d/9da50e155df72bce55cb69f51f1dbb4b62d23740fb99f6178bb27f22ebdf8a46.tar.gz
==> Moving resource stage
        source: /tmp/csbcfmpd/spack-stage/resource-cpanm-2tubazheh7m4pdei3eugarkkduubhga6/spack-src/
        destination: /tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src/cpanm/cpanm
==> Applied patch /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack/var/spack/repos/builtin/packages/perl/zlib-ng.patch
==> Ran patch() for perl
==> perl: Executing phase: 'configure'
==> perl: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j1'

1 error found in build log:
     2772       -lpthread   \
     2773    
     2774    chmod 755 ../../lib/auto/threads/shared/shared.so
     2775    make[1]: Leaving directory '/tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src/dist/threads-shared'
     2776    LD_LIBRARY_PATH=/tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src  ./perl -Ilib -I. -f pod/buildtoc -q
     2777    Cwd.c: loadable library and perl binaries are mismatched (got handshake key 0xf380080, needed 0xcd00080)
  >> 2778    make: *** [makefile:426: pod/perltoc.pod] Error 1
```

--- a/lib/spack/spack/build_environment.py
+++ b/lib/spack/spack/build_environment.py
@@ -441,7 +441,7 @@ class FilterDefaultDynamicLinkerSearchPaths:
     """Remove rpaths to directories that are default search paths of the dynamic linker."""
 
     def __init__(self, dynamic_linker: Optional[str]) -> None:
-        # Idenitify directories by (inode, device) tuple, which handles symlinks too.
+        # Identify directories by (inode, device) tuple, which handles symlinks too.
         self.default_path_identifiers: Set[Tuple[int, int]] = set()
         if not dynamic_linker:
             return

--- a/lib/spack/spack/build_environment.py
+++ b/lib/spack/spack/build_environment.py
@@ -37,13 +37,14 @@ import io
 import multiprocessing
 import os
 import re
+import stat
 import sys
 import traceback
 import types
 from collections import defaultdict
 from enum import Flag, auto
 from itertools import chain
-from typing import Dict, List, Set, Tuple
+from typing import Dict, List, Optional, Set, Tuple
 
 import archspec.cpu
 
@@ -74,6 +75,7 @@ import spack.stage
 import spack.store
 import spack.subprocess_context
 import spack.util.executable
+import spack.util.libc
 from spack import traverse
 from spack.context import Context
 from spack.error import InstallError, NoHeadersError, NoLibrariesError
@@ -435,6 +437,35 @@ def optimization_flags(compiler, target):
     return result
 
 
+class FilterDefaultDynamicLinkerSearchPaths:
+    """Remove rpaths to directories that are default search paths of the dynamic linker."""
+
+    def __init__(self, dynamic_linker: Optional[str]) -> None:
+        # Idenitify directories by (inode, device) tuple, which handles symlinks too.
+        self.default_path_identifiers: Set[Tuple[int, int]] = set()
+        if not dynamic_linker:
+            return
+        for path in spack.util.libc.default_search_paths_from_dynamic_linker(dynamic_linker):
+            try:
+                s = os.stat(path)
+                if stat.S_ISDIR(s.st_mode):
+                    self.default_path_identifiers.add((s.st_ino, s.st_dev))
+            except OSError:
+                continue
+
+    def is_dynamic_loader_default_path(self, p: str) -> bool:
+        try:
+            s = os.stat(p)
+            return (s.st_ino, s.st_dev) in self.default_path_identifiers
+        except OSError:
+            return False
+
+    def __call__(self, dirs: List[str]) -> List[str]:
+        if not self.default_path_identifiers:
+            return dirs
+        return [p for p in dirs if not self.is_dynamic_loader_default_path(p)]
+
+
 def set_wrapper_variables(pkg, env):
     """Set environment variables used by the Spack compiler wrapper (which have the prefix
     `SPACK_`) and also add the compiler wrappers to PATH.
@@ -508,17 +539,17 @@ def set_wrapper_variables(pkg, env):
     def update_compiler_args_for_dep(dep):
         if dep in link_deps and (not is_system_path(dep.prefix)):
             query = pkg.spec[dep.name]
-            dep_link_dirs = list()
+            dep_link_dirs = []
             try:
                 # In some circumstances (particularly for externals) finding
                 # libraries packages can be time consuming, so indicate that
                 # we are performing this operation (and also report when it
                 # finishes).
-                tty.debug("Collecting libraries for {0}".format(dep.name))
+                tty.debug(f"Collecting libraries for {dep.name}")
                 dep_link_dirs.extend(query.libs.directories)
-                tty.debug("Libraries for {0} have been collected.".format(dep.name))
+                tty.debug(f"Libraries for {dep.name} have been collected.")
             except NoLibrariesError:
-                tty.debug("No libraries found for {0}".format(dep.name))
+                tty.debug(f"No libraries found for {dep.name}")
 
             for default_lib_dir in ["lib", "lib64"]:
                 default_lib_prefix = os.path.join(dep.prefix, default_lib_dir)
@@ -532,7 +563,7 @@ def set_wrapper_variables(pkg, env):
             try:
                 _prepend_all(include_dirs, query.headers.directories)
             except NoHeadersError:
-                tty.debug("No headers found for {0}".format(dep.name))
+                tty.debug(f"No headers found for {dep.name}")
 
     for dspec in pkg.spec.traverse(root=False, order="post"):
         if dspec.external:
@@ -552,9 +583,18 @@ def set_wrapper_variables(pkg, env):
         lib_path = os.path.join(pkg.prefix, libdir)
         rpath_dirs.insert(0, lib_path)
 
+    filter_default_dynamic_linker_search_paths = FilterDefaultDynamicLinkerSearchPaths(
+        pkg.compiler.default_dynamic_linker
+    )
+
     link_dirs = list(dedupe(filter_system_paths(link_dirs)))
     include_dirs = list(dedupe(filter_system_paths(include_dirs)))
     rpath_dirs = list(dedupe(filter_system_paths(rpath_dirs)))
+    rpath_dirs = filter_default_dynamic_linker_search_paths(rpath_dirs)
+
+    implicit_rpaths = filter_default_dynamic_linker_search_paths(pkg.compiler.implicit_rpaths())
+    if implicit_rpaths:
+        env.set("SPACK_COMPILER_IMPLICIT_RPATHS", ":".join(implicit_rpaths))
 
     # Spack managed directories include the stage, store and upstream stores. We extend this with
     # their real paths to make it more robust (e.g. /tmp vs /private/tmp on macOS).
@@ -840,10 +880,6 @@ def setup_package(pkg, dirty, context: Context = Context.BUILD):
             load_module(mod)
 
     load_external_modules(pkg)
-
-    implicit_rpaths = pkg.compiler.implicit_rpaths()
-    if implicit_rpaths:
-        env_mods.set("SPACK_COMPILER_IMPLICIT_RPATHS", ":".join(implicit_rpaths))
 
     # Make sure nothing's strange about the Spack environment.
     validate(env_mods, tty.warn)

--- a/lib/spack/spack/compiler.py
+++ b/lib/spack/spack/compiler.py
@@ -415,14 +415,19 @@ class Compiler:
         return list(paths_containing_libs(link_dirs, all_required_libs))
 
     @property
-    def default_libc(self) -> Optional["spack.spec.Spec"]:
-        """Determine libc targeted by the compiler from link line"""
+    def default_dynamic_linker(self) -> Optional[str]:
+        """Determine default dynamic linker from compiler link line"""
         output = self.compiler_verbose_output
 
         if not output:
             return None
 
-        dynamic_linker = spack.util.libc.parse_dynamic_linker(output)
+        return spack.util.libc.parse_dynamic_linker(output)
+
+    @property
+    def default_libc(self) -> Optional["spack.spec.Spec"]:
+        """Determine libc targeted by the compiler from link line"""
+        dynamic_linker = self.default_dynamic_linker
 
         if not dynamic_linker:
             return None

--- a/lib/spack/spack/util/libc.py
+++ b/lib/spack/spack/util/libc.py
@@ -9,7 +9,7 @@ import re
 import shlex
 import sys
 from subprocess import PIPE, run
-from typing import Optional
+from typing import List, Optional
 
 import spack.spec
 import spack.util.elf
@@ -32,6 +32,22 @@ def _libc_from_ldd(ldd: str) -> Optional["spack.spec.Spec"]:
         return spack.spec.Spec(f"glibc@={version_str.group(1)}")
     except Exception:
         return None
+
+
+def default_search_paths_from_dynamic_linker(dynamic_linker: str) -> List[str]:
+    """If the dynamic linker is glibc at a certain version, we can query the hard-coded library
+    search paths"""
+    try:
+        result = run([dynamic_linker, "--help"], stdout=PIPE, stderr=PIPE, check=False)
+        assert result.returncode == 0
+        out = result.stdout.decode("utf-8")
+    except Exception:
+        return []
+
+    return [
+        match.group(1).strip()
+        for match in re.finditer(r"^  (/.+) \(system search path\)$", out, re.MULTILINE)
+    ]
 
 
 def libc_from_dynamic_linker(dynamic_linker: str) -> Optional["spack.spec.Spec"]:

